### DISCUSSION TITLE
Script linking Interface and IPAddress

### DISCUSSION THREAD

**Discussion by jscooksey** [netbox-community/netbox#8986] [Q&A] [answered]
Using scripting I'm having some trouble figuring out how to link an Interface to an IPAddress.

I'm creating an Interface
```python
new_int = Interface(
    device=device,
    name=f'Loopback{interface_id}',
    description=description,
    type=InterfaceTypeChoices.TYPE_VIRTUAL,
    enabled=True
)
new_int.save()
```

I'm creating an IPAddress
```python
new_ip = IPAddress(
    address=str(core_router_ip) + "/30",
    vrf=vrf,
    tenant=tenant,
    role=IPAddressRoleChoices.ROLE_LOOPBACK,
    status=IPAddressStatusChoices.STATUS_ACTIVE,
    description=description,
)
new_ip.save()
```

I've tried adding 'interface=new_int,' into the arguments to create IPAddress

I've tried:

```python
new_ip.interface.add(new_nbn_int)
new_ip.interface.only(new_nbn_int)
```

It'll be something I'll kick myself over I'm sure, and I need to do some Django training.

Thanks,

**Comment by jeremystretch** [comment, marked as answer]
When you're not sure how to interact with an object, first reference the model's source code: More of than not it will have all the clues you need. You can find the source for the IPAddress model in [`ipam/models/ip.py`](https://github.com/netbox-community/netbox/blob/894665b067616cb7b29c4522460f55daf53f3e61/netbox/ipam/models/ip.py#L763).

You can see that there's no `interface` field on the IPAddress model.* Instead, we see a generic foreign key named `assigned_object`. This field can be set to assign an IP address to a device interface, a VM interface, or an FHRP group.

```
new_ip.assigned_object = new_int
new_ip.save()
```

Hope that helps!

*IIRC there was an `interface` field on the IPAddress model some time ago, but it was replaced with a generic foreign key when we split device and VM interfaces into separate models (whenever that was).

**Reply by jscooksey** [reply]
Thanks @jeremystretch  I'll do more digging through the Netbox code as I continue to build.

**Reply by candlerb** [reply]
There's a script here which may be helpful:
https://github.com/netbox-community/reports/blob/master/scripts/create_vm.py

BEWARE: don't just do `new_ip.save()`: first call `new_ip.full_clean()`.  If you don't do this, then Django ORM will allow you to save invalid data!
